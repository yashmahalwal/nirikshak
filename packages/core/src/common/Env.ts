/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
