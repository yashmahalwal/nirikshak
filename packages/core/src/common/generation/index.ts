export { generateFaker } from "./fakerGen";
export { generateCustom } from "./customGen";
