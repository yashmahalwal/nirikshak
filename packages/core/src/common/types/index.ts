export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
