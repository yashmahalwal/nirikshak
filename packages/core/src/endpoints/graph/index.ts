export { traverseGraph } from "./traverseGraph";
export {
  serializeNodeName,
  parseNodeName,
  ParsedNode,
  NodeEntry,
  NodeMap,
  NodeName,
} from "./nodeTypes";
export { makeGraph } from "./makeGraph";
export { generateNodes } from "./generateNodes";
