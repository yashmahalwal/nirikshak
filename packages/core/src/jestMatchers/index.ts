/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-namespace */
import { toMatchStatus } from "./statusMatchers";
import { toMatchHeaders } from "./headerMatchers";
import { toMatchBody } from "./bodyMatchers";

expect.extend({
    toMatchStatus,
    toMatchHeaders,
    toMatchBody,
});

declare global {
    namespace jest {
        interface Matchers<R> {
            toMatchStatus(expected: number[]): R;
            toMatchHeaders(expected: boolean): R;
            toMatchBody(expected: boolean): R;
        }
    }
}
