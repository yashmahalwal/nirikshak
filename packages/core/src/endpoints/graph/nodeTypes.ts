import { URLString, isURLString } from "../types/urlString";
import { MethodType, Cases } from "..";
import { isMethodType, isCaseType,  } from "../types/helpers";
import { InputSemantics, InputBodies } from "../types/input";
import { OutputSemantics, OutputBodies } from "../types/output";

export type NodeName = string;

export function isNodeName(str: any): str is NodeName {
    if (typeof str !== "string") return false;

    const parts = str.split(";;");
    return (
        parts.length === 4 &&
        isMethodType(parts[0]) &&
        !isNaN(parseInt(parts[1])) &&
        isURLString(parts[2]) &&
        isCaseType(parts[3])
    );
}

export interface NodeEntry {
    url: URLString;
    input: InputSemantics & Partial<InputBodies>;
    output: (OutputSemantics & Partial<OutputBodies>)[];
}

export interface ParsedNode {
    method: MethodType;
    methodIndex: number;
    url: URLString;
    caseValue: Cases;
}

export function serializeNodeName(
    method: MethodType,
    methodIndex: number,
    url: URLString,
    caseValue: Cases
): NodeName {
    return `${method};;${methodIndex};;${url};;${caseValue}`;
}

export function parseNodeName(str: NodeName): ParsedNode {
    const parts = str.split(";;");
    return {
        method: parts[0],
        methodIndex: parseInt(parts[1]),
        url: parts[2],
        caseValue: parts[3],
    } as ParsedNode;
}

export type NodeMap = Map<NodeName, NodeEntry>;
