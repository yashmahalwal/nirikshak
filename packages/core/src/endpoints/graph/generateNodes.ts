import { Description, MethodType, Cases } from "../types";
import { NodeMap, serializeNodeName } from "./nodeTypes";
import { OutputSemantics, OutputBodies, Outputs } from "../types/output";

export function generateNodes(description: Description): NodeMap {
    const nodeMap: NodeMap = new Map();
    for (const method in description) {
        const entry = description[method as MethodType];
        const methodArr = Array.isArray(entry) ? entry : [entry];

        for (let i = 0; i < methodArr.length; i++) {
            const methodEntry = methodArr[i];
            for (const caseValue in methodEntry?.output) {
                const caseArr = (Array.isArray(
                    methodEntry!.output[caseValue as keyof Outputs[MethodType]]
                )
                    ? methodEntry!.output[
                          caseValue as keyof Outputs[MethodType]
                      ]
                    : [
                          methodEntry!.output[
                              caseValue as keyof Outputs[MethodType]
                          ],
                      ]) as (OutputSemantics & Partial<OutputBodies>)[];
                nodeMap.set(
                    serializeNodeName(
                        method as MethodType,
                        i,
                        methodEntry!.url,
                        caseValue as Cases
                    ),
                    {
                        input: methodEntry!.input,
                        url: methodEntry!.url,
                        output: caseArr,
                    }
                );
            }
        }
    }

    return nodeMap;
}
