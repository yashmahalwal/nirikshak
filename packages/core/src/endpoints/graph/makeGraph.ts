import { NodeMap, NodeName, parseNodeName } from "./nodeTypes";
import Graph from "graph-data-structure";

export function isNeighbour(source: NodeName, target: NodeName): boolean {
    const { method: sourceMethod, caseValue: sourceCase } = parseNodeName(
        source
    );
    const { method: targetMethod, caseValue: targetCase } = parseNodeName(
        target
    );

    switch (sourceMethod) {
        case "GET":
            switch (sourceCase) {
                case "POSITIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "POSITIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
                case "NEGATIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "NEGATIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
            }
        // eslint-disable-next-line no-fallthrough
        case "DELETE":
            switch (sourceCase) {
                case "POSITIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "NEGATIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
                case "NEGATIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "NEGATIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
            }
        // eslint-disable-next-line no-fallthrough
        case "PUT":
            switch (sourceCase) {
                case "POSITIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "POSITIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
                default:
                    return false;
            }
        case "POST":
            switch (sourceCase) {
                case "POSITIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "POSITIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
                case "NEGATIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "POSITIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
                default:
                    return false;
            }
        case "PATCH":
            switch (sourceCase) {
                case "POSITIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "POSITIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
                case "NEGATIVE":
                    if (
                        (targetMethod === "GET" && targetCase === "NEGATIVE") ||
                        (targetMethod === "DELETE" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" && targetCase === "POSITIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "POSITIVE") ||
                        (targetMethod === "PATCH" &&
                            targetCase === "NEGATIVE") ||
                        (targetMethod === "PUT" &&
                            targetCase === "DESTRUCTIVE") ||
                        (targetMethod === "POST" &&
                            targetCase === "DESTRUCTIVE")
                    )
                        return true;
                    return false;
                default:
                    return false;
            }
    }
}

export function makeGraph(nodeMap: NodeMap): ReturnType<typeof Graph> {
    const graph = Graph();

    const keys = Array.from(nodeMap.keys());
    for (const source of keys) {
        for (const target of keys) {
            graph.addNode(source);
            graph.addNode(target);
            graph.addEdge(source, target, isNeighbour(source, target) ? 1 : 0);
        }
    }

    return graph;
}
