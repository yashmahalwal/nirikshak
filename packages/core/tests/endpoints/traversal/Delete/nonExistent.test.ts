import getPort from "get-port";
import { createServer } from "http";
import {
    Resource,
    SchemaHelpers,
    Collection,
    generateResource,
    makeRequest,
} from "../../../../src";
import { RANDOMNESS_ITERATIONS } from "../../../../src/common/Env";
import supertest from "supertest";
import {
    NodeMap,
    parseNodeName,
} from "../../../../src/endpoints/graph/nodeTypes";
import { generateNodes } from "../../../../src/endpoints/graph/generateNodes";
import "../../../../src/jestMatchers/index";

const resourceSchema: Resource = {
    id: "faker:random.uuid",
    name: ["faker:name.firstName", "faker:name.lastName"],
    age: "faker:random.number",
    address: {
        types: [{ type: "faker:zipCode", plural: true }],
        fields: [
            {
                field: {
                    zipCode: "faker:addresss.zipCode",
                    city: "faker:address.city",
                },
                plural: true,
            },
        ],
    },
    bloodGroup: {
        type: { function: "custom:bloodGroup", args: ["O+"] },
        optional: true,
    },
};

const schemaHelpers: SchemaHelpers = {
    async bloodGroup(preferred?: string) {
        return preferred ?? "A+";
    },
};

const description = {
    DELETE: {
        input: {
            semantics: {},
        },
        url: "/student/{resource:id}",
        output: {
            POSITIVE: {
                semantics: {
                    status: 200,
                    headers: {
                        "content-type": "application/json",
                        "item-id": "resource:id",
                    },
                },
                body: "custom:check",
            },
        },
    },
};
const collection: Collection = new Map();
const server = createServer((req, res) => {
    if (req.method === "DELETE" && req.url?.startsWith("/student")) {
        const [, , id, query] = req.url.split(/\?|\//);
        res.setHeader("item-id", id);
        if (collection.has(id)) {
            res.statusCode = 200;
            res.setHeader("content-type", "application/json");
            const { callergroup: group } = req.headers;
            const o: Record<string, any> = {
                studentAddresses: collection.get(id)!.address,
            };

            if (query) {
                const [key, value] = query.split("=");
                o.query = {
                    [key]: value,
                };
            }

            if (group) {
                res.setHeader("callerGroup", group);
            }
            res.write(JSON.stringify(o));
        } else {
            res.statusCode = 404;
        }
        res.end();
    }
});

describe(`Deleting already deleted resource`, () => {
    beforeAll(async (done) => {
        for (let i = 0; i < RANDOMNESS_ITERATIONS; i++) {
            const newResource = await generateResource(
                resourceSchema,
                schemaHelpers
            );
            collection.set(newResource.id, newResource);
        }

        server.listen(await getPort(), done);
    });
    const nodeMap: NodeMap = generateNodes(description);

    for (const [key, entry] of nodeMap.entries())
        test(key, async () => {
            expect.hasAssertions();
            const parsedNode = parseNodeName(key);
            const app = supertest(server);

            try {
                await makeRequest(
                    entry.url,
                    parsedNode.method,
                    parsedNode.caseValue,
                    app,
                    entry.input,
                    await generateResource(resourceSchema, schemaHelpers),
                    schemaHelpers,
                    collection
                );
            } catch (e) {
                expect(e).toMatchSnapshot();
            }
        });
    afterAll((done) => server.close(done));
});
