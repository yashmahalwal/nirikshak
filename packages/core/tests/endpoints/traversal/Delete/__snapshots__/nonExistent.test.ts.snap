// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deleting already deleted resource DELETE;;0;;/student/{resource:id};;POSITIVE 1`] = `[Error: Cannot delete non existent resource from collection]`;
